File: day4.py
#!/usr/bin/env python3
from utils import runner
import re

def part1(lines):
    keys = set(['byr','iyr','eyr','hgt','hcl','ecl', 'pid'])
    count = 0
    current = set()
    for line in lines:
        if line == "":
            if current.issuperset(keys):
                count +=  1
            current = set()
        else:
            for pair in line.split(" "):
                current.add(pair.split(':')[0])
    return count

def part2(lines):
    hcl_p = re.compile("^#[0-9a-f]{6}$")
    pid_p = re.compile("^[0-9]{9}$")
    rules = {'byr': lambda v: len(v) == 4 and int(v) >= 1920 and int(v) <= 2002,
            'iyr': lambda v: len(v) == 4 and int(v) >= 2010 and int(v) <= 2020,
            'eyr': lambda v: len(v) == 4 and int(v) >= 2020 and int(v) <= 2030,
            'hgt': lambda v: len(v) >=3 and ((v[-2:] == 'cm' and v[:-2].isnumeric and int(v[:-2]) >= 150 and int(v[:-2]) <= 193) or \
                                             (v[-2:] == 'in' and v[:-2].isnumeric and int(v[:-2]) >= 59 and int(v[:-2]) <= 76)),
            'hcl': lambda v: hcl_p.match(v),
            'ecl': lambda v: v in 'amb blu brn gry grn hzl oth'.split(' '),
            'pid': lambda v: pid_p.match(v)}

    count = 0
    current = {}
    for line in lines:
        if line == "":
            correct = 0
            if len(list(filter(lambda pair: pair[0] in rules and rules[pair[0]](pair[1]), [(k, v) for k, v in current.items()]))) == 7:
                count +=  1
            current = {}
        else:
            pairs = line.split(" ")
            for pair in pairs:
                parts = pair.split(':')
                current[parts[0]] = parts[1]
    return count

if __name__ == "__main__":
    with open("4s.txt") as f:
        sample1 = [line.strip() for line in f.readlines()]

    with open("42i.txt") as f:
        sample2 = [line.strip() for line in f.readlines()]

    with open("42v.txt") as f:
        sample3 = [line.strip() for line in f.readlines()]

    with open("day4.txt") as f:
        lines = [line.strip() for line in f.readlines()]

    runner(part1, sample1, 2, lines, 235)
    runner(part2, sample2, 0, lines, 194)
    runner(part2, sample3, 4, lines, 194)

File: src/day4.rs
use std::collections::HashMap;
use std::collections::HashSet;
use std::fs;
use regex::Regex;

fn part1(lines: &[&str]) -> i32 {
    let keys: HashSet<&str> = ["byr","iyr","eyr","hgt","hcl","ecl", "pid"].iter().cloned().collect();
    let mut count = 0;
    let mut current = HashSet::new();
    for line in lines {
        if line.is_empty() {
            if current.is_superset(&keys) {
                count += 1;
            }
            current = HashSet::new();
        }
        else {
            for pair in line.split(' ') {
                current.insert(pair.split(':').next().unwrap());
            }
        }
    }
    count
}

fn part2(lines: &[&str]) -> i32 {
    let hcl_p = Regex::new(r"^#[0-9a-f]{6}$").unwrap();
    let pid_p = Regex::new(r"^[0-9]{9}$").unwrap();
    let mut count = 0;
    let mut current: HashMap<&str, &str> = HashMap::new();
    for line in lines {
        if line.is_empty() {
            let mut n = 0;
            for (k, v) in current {
                match k {
                    "byr" => if v.len() == 4 && v.parse::<usize>().unwrap() >= 1920 && v.parse::<usize>().unwrap() <= 2002 { n += 1; },
                    "iyr" => if v.len() == 4 && v.parse::<usize>().unwrap() >= 2010 && v.parse::<usize>().unwrap() <= 2020 { n += 1; },
                    "eyr" => if v.len() == 4 && v.parse::<usize>().unwrap() >= 2020 && v.parse::<usize>().unwrap() <= 2030 { n += 1; },
                    "ecl" => if "amb blu brn gry grn hzl oth".split(' ').any(|x| x == v) { n += 1; },
                    "hcl" => if hcl_p.is_match(v) { n += 1 },
                    "pid" => if pid_p.is_match(v) { n += 1 },
                    "hgt" => if v.len() >= 3 && (
                        (&v[v.len() - 2..v.len()] == "cm" &&
                            v[0..v.len() - 2].parse::<usize>().unwrap() >= 150 &&
                            v[0..v.len() - 2].parse::<usize>().unwrap() <= 193) ||
                        (&v[v.len() - 2..v.len()] == "in" &&
                            v[0..v.len() - 2].parse::<usize>().unwrap() >= 59 &&
                            v[0..v.len() - 2].parse::<usize>().unwrap() <= 76)) { n += 1; },
                    _ => ()
                }
            }
            if n == 7 {
                count += 1;
            }

            current = HashMap::new();
        }
        else {
            for pair in line.split(' ') {
                let k = pair.split(':').next().unwrap();
                let v = pair.split(':').nth(1).unwrap();
                current.insert(k, v);
            }
        }
    }
    count
}

fn runner(func: &dyn Fn(&[&str]) -> i32, sample: &[&str], expected: i32, input: &[&str], answer: Option<i32>) {
    assert_eq!(func(sample), expected);
    let ans = func(input);
    println!("{}", ans);

    if let Some(x) = answer {
        assert_eq!(ans, x);
    }
}

pub fn day4() {
    let contents = fs::read_to_string("day4.txt").unwrap();
    let input: Vec<&str> = contents.lines().collect();

    let sample_contents = fs::read_to_string("4s.txt").unwrap();
    let sample: Vec<&str> = sample_contents.lines().collect();

    let sample_contentsi = fs::read_to_string("42i.txt").unwrap();
    let samplei: Vec<&str> = sample_contentsi.lines().collect();

    let sample_contentsv = fs::read_to_string("42v.txt").unwrap();
    let samplev: Vec<&str> = sample_contentsv.lines().collect();

    runner(&part1, &sample, 2, &input, Some(235));
    runner(&part2, &samplei, 0, &input, Some(194));
    runner(&part2, &samplev, 4, &input, Some(194));
}